### PR TITLE
add application.mergeFormUrlAsStruct LDEV-3113

### PR DESCRIPTION
**Pull request by zspitzer**
- disables the merging of dot notation url or form keys into structs, matches ACF behaviour
- does change the lucee loader signature :(

https://luceeserver.atlassian.net/browse/LDEV-3113

### CODE REVIEW

File: core/src/main/java/lucee/runtime/functions/system/GetApplicationSettings.java [hunk 1]
```diff
@@ -135,6 +135,8 @@ public static Struct call(PageContext pc, boolean suppressFunctions) throws Page
 		sct.setEL("triggerDataMember", Caster.toBoolean(ac.getTriggerComponentDataMember()));
 		sct.setEL("sameformfieldsasarray", Caster.toBoolean(ac.getSameFieldAsArray(Scope.SCOPE_FORM)));
 		sct.setEL("sameurlfieldsasarray", Caster.toBoolean(ac.getSameFieldAsArray(Scope.SCOPE_URL)));
+		sct.setEL("mmergeFormUrlAsStruct", Caster.toBoolean(ac.getMmergeFormUrlAsStruct()));
```
Comment (Vintic) [inline core/src/main/java/lucee/runtime/functions/system/GetApplicationSettings.java]: why "mmergeFormUrlAsStruct" but not "mergeFormUrlAsStruct"?

Comment (zspitzer) [reply core/src/main/java/lucee/runtime/functions/system/GetApplicationSettings.java]: good catch :)

File: core/src/main/java/lucee/runtime/listener/ClassicApplicationContext.java [hunk 2]
```diff
@@ -807,6 +807,11 @@ public boolean getSameFieldAsArray(int scope) {
 		return b.booleanValue();
 	}
 
+	@Override
+	public boolean getMmergeFormUrlAsStruct(){
```
Comment (Vintic) [inline core/src/main/java/lucee/runtime/listener/ClassicApplicationContext.java]: Same here

File: core/src/main/java/lucee/runtime/listener/ModernApplicationContext.java [hunk 3]
```diff
@@ -238,6 +239,7 @@
 	private boolean initSetDomainCookies;
 	private boolean initSetSessionManagement;
 	private boolean initScriptProtect;
+	private boolean initMmergeFormUrlAsStruct;
```
Comment (Vintic) [inline core/src/main/java/lucee/runtime/listener/ModernApplicationContext.java]: +

File: core/src/main/java/lucee/runtime/type/scope/URLImpl.java [hunk 4]
```diff
@@ -57,21 +58,22 @@ public void setEncoding(ApplicationContext ac, String encoding) throws Unsupport
...
 	}
 
 	@Override
 	public void initialize(PageContext pc) {
 		if (encoding == null) encoding = pc.getWebCharset().name();
 		if (scriptProtected == ScriptProtected.UNDEFINED) {
 			scriptProtected = ((pc.getApplicationContext().getScriptProtect() & ApplicationContext.SCRIPT_PROTECT_URL) > 0) ? ScriptProtected.YES : ScriptProtected.NO;
-		}
-
+			structMerge = pc.getApplicationContext().getMmergeFormUrlAsStruct();
```
Comment (Vintic) [inline core/src/main/java/lucee/runtime/type/scope/URLImpl.java]: +

File: core/src/main/java/lucee/runtime/type/util/KeyConstants.java [hunk 5]
```diff
@@ -1018,6 +1018,7 @@
 	public static final Key _Documentation = KeyImpl._const("Documentation");
 	public static final Key _sameformfieldsasarray = KeyImpl._const("sameformfieldsasarray");
 	public static final Key _sameurlfieldsasarray = KeyImpl._const("sameurlfieldsasarray");
+	public static final Key _mmergeFormUrlAsStruct = KeyImpl._const("mmergeFormUrlAsStruct");
```
Comment (Vintic) [inline core/src/main/java/lucee/runtime/type/util/KeyConstants.java]: +

File: loader/src/main/java/lucee/runtime/listener/ApplicationContext.java [hunk 6]
```diff
@@ -235,6 +235,8 @@
 	 */
 	public boolean getSameFieldAsArray(int scope);
 
+	public boolean getMmergeFormUrlAsStruct();
```
Comment (Vintic) [inline loader/src/main/java/lucee/runtime/listener/ApplicationContext.java]: sorry for that

File: core/src/main/java/lucee/runtime/type/scope/URLImpl.java [hunk 7]
```diff
@@ -91,8 +93,9 @@ public void reinitialize(ApplicationContext ac) {
 			if (scriptProtected == ScriptProtected.UNDEFINED) {
 				scriptProtected = ((ac.getScriptProtect() & ApplicationContext.SCRIPT_PROTECT_URL) > 0) ? ScriptProtected.YES : ScriptProtected.NO;
 			}
+			structMerge = ac.getMmergeFormUrlAsStruct();
```
Comment (Vintic) [inline core/src/main/java/lucee/runtime/type/scope/URLImpl.java]: +

File: core/src/main/java/lucee/runtime/functions/system/GetApplicationSettings.java [hunk 8]
```diff
@@ -135,6 +135,8 @@ public static Struct call(PageContext pc, boolean suppressFunctions) throws Page
 		sct.setEL("triggerDataMember", Caster.toBoolean(ac.getTriggerComponentDataMember()));
 		sct.setEL("sameformfieldsasarray", Caster.toBoolean(ac.getSameFieldAsArray(Scope.SCOPE_FORM)));
 		sct.setEL("sameurlfieldsasarray", Caster.toBoolean(ac.getSameFieldAsArray(Scope.SCOPE_URL)));
+		sct.setEL("mergeFormUrlAsStruct", Caster.toBoolean(ac.getMmergeFormUrlAsStruct()));
```
Comment (Vintic) [inline core/src/main/java/lucee/runtime/functions/system/GetApplicationSettings.java]: +

**Review by Vintic — COMMENTED** [review]
ignore please

**Comment by zspitzer** [comment]
> ignore please

no worries, i did when i couldn't sleep one night, thanks for picking up my mistakes :)

**Comment by michaeloffner** [comment]
@zspitzer we need to add this also to the tag cfapplication, all features we add to Application.cfc we also do with the tag cfapplication. The reason is that the feature should be available if you do 
<cfapplication action="update" ...>

**Comment by zspitzer** [comment]
@michaeloffner ah, that's coz i was following the `sameurlfieldsasarray` and `sameformfieldsasarray` which are also missing from https://github.com/lucee/Lucee/blob/5.3/core/src/main/java/lucee/runtime/tag/Application.java

I noticed that the existing implementation those parameters can trigger re-processing the form and url scope

how do I avoid changing the loader signature?

**Comment by michaeloffner** [comment]
i have adapted the PR manually with soem changes, more details in the ticket.

**Pull request closed without merging**